### PR TITLE
[CLI] add support for listing apps

### PR DESCRIPTION
**Pull request by nicolai86**
## What does this PR do?

This PR changes the CLI to add support for listing apps. For our launch of BYOC support we want to be able to list apps and also filter by cluster which executes the application.

### listing apps

```
❯ lightning list apps
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ id                         ┃ name                       ┃ status  ┃ cluster       ┃ created    ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ 01g9g4p2my8a3na6jx4kdxtmzn │ lightning/notebook-manager │ running │ litng-ai-03   │ 2022-08-02 │
│ 01g92nrr7r06zh8nhddzkhg213 │ fresh-boyd-193             │ stopped │ test-raphael9 │ 2022-07-28 │
└────────────────────────────┴────────────────────────────┴─────────┴───────────────┴────────────┘
```

the `--cluster-id` flag can be passed to list only apps in a specific cluster:

```
❯ lightning list apps  --cluster-id test-raphael9
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ id                         ┃ name           ┃ status  ┃ cluster       ┃ created    ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ 01g92nrr7r06zh8nhddzkhg213 │ fresh-boyd-193 │ stopped │ test-raphael9 │ 2022-07-28 │
└────────────────────────────┴────────────────┴─────────┴───────────────┴────────────┘
```

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @tchaton @rohitgr7

### CODE REVIEW

File: src/lightning_app/cli/cmd_apps.py
Comment (alecmerdler) [inline src/lightning_app/cli/cmd_apps.py:20]: It would be nice to have a `output` option, which could be `table|json|yaml|text` for machine readable output.

**Review by alecmerdler — APPROVED** [review]
Nice job!

File: src/lightning_app/cli/cmd_apps.py
Comment (nicolai86) [reply src/lightning_app/cli/cmd_apps.py:20]: I agree. Let's introduce that in a separate PR though, as it should be honored in all CLI pieces that output text.

File: src/lightning_app/cli/lightning_cli_list.py
Comment (awaelchli) [inline src/lightning_app/cli/lightning_cli_list.py:29]: Is the naming already decided here?

In our vision to [design a CLI](https://www.notion.so/lightningai/CLI-Redesign-92270473ae494a36b0f59db3e325b1d4) for both Lightning and Grid (eventually becoming one and the same), we imagined having a single top level command to retrieve information about resources, called `show`.

I'm ok having a list command, but want to make sure we know how it is going to fit into the larger picture and what else it is going to do in the future.

cc @adam-lightning

Comment (nicolai86) [reply src/lightning_app/cli/lightning_cli_list.py:29]: this is per the BYOC CLI RFC that was reviewed and approved from the right places (see grid-rfcs # 35). The design you linked is for grid, and as far as I know for lightning CLI folks wanted a different approach.

Comment (nicolai86) [reply src/lightning_app/cli/lightning_cli_list.py:29]: reading the RFC this only pertains to showing individual apps, which is different from listing all the apps that are currently running. We will have a different command to fetch details about an individual app.

Comment (adam-lightning) [reply src/lightning_app/cli/lightning_cli_list.py:29]: I recall some time ago parts of the discussion about `list` and `show` commands. But don't really know if and where it's documented... 😓 
My take is I'm ok with separate `show` and `list` as long as their intended use is clear i.e. `list` just lists all resources of some type and gives enough information to later "inspect" or `show` more info by using dedicated command.

Comment (tchaton) [reply src/lightning_app/cli/lightning_cli_list.py:29]: @awaelchli You should have the last call on the CLI API design there. How do you want to proceed ?

**Review by Borda — APPROVED** [review]

File: src/lightning_app/cli/cmd_apps.py
Comment (hhsecond) [inline src/lightning_app/cli/cmd_apps.py:31]: nit: we could probably use a higher limit as well here (regardless, both are not apparently implemented in the BE)

File: src/lightning_app/cli/cmd_apps.py
Comment (hhsecond) [inline src/lightning_app/cli/cmd_apps.py:48]: nit: we should probably do a bit better than just one to one mapping, like the "terminating" one we did below for other statuses, like "restarting". Probably looking at the FE codebase is a good reference?

File: src/lightning_app/cli/cmd_apps.py
Comment (hhsecond) [inline src/lightning_app/cli/cmd_apps.py:29]: we should probably manager exceptions here. Even if not exhaustive, atleast basic ones like cluster permission denied etc - super useful when users make spelling mistakes on cluster id

**Review by hhsecond — COMMENTED** [review]
Nice job! Some minor change requests but since it's merging as "experimental", I don't want to block the PR. The exception handling is something I think we should do now though, but again if you are keeping it for later, please track that in a ticket.

**Review by hhsecond — APPROVED** [review]
Approving to unblock as I am signing off

File: src/lightning_app/cli/cmd_apps.py
Comment (nicolai86) [reply src/lightning_app/cli/cmd_apps.py:31]: are you talking about the page size?

Comment (hhsecond) [reply src/lightning_app/cli/cmd_apps.py:31]: yes!

File: src/lightning_app/cli/cmd_apps.py
Comment (nicolai86) [reply src/lightning_app/cli/cmd_apps.py:29]: I'll check what typed exceptions we have in our OpenAPI model today and see if I can catch some meaningful errors here :+1:

File: src/lightning_app/cli/cmd_apps.py
Comment (nicolai86) [reply src/lightning_app/cli/cmd_apps.py:48]: let me see what other transitions we have in there and add them below :+1:

File: src/lightning_app/cli/cmd_apps.py
Comment (nicolai86) [reply src/lightning_app/cli/cmd_apps.py:14]: It's an abstraction I put into place so that I don't put too much logic into the individual CLI command - pretty much like `_run_app`, just as a class.

I'll make it protected :+1:

File: src/lightning_app/cli/cmd_apps.py
Comment (nicolai86) [reply src/lightning_app/cli/cmd_apps.py:31]: made explicit and bumbed to 100 for now :+1:

File: src/lightning_app/cli/cmd_apps.py
Comment (nicolai86) [reply src/lightning_app/cli/cmd_apps.py:30]: just did some validation on the API - unknown clusters just return an empty response with no error. The API can only return two errors: permission denied when you don't have access to the project, and internal error if anything in the service failed.

because we don't expose project management today, the only way to select a project is via `LIGHTNING_CLOUD_PROJECT_ID` environment variable, in which case `_get_project` already provides a useful error message: 

```
ValueError: Environment variable LIGHTNING_CLOUD_PROJECT_ID is set but could not find an associated project.
```

Let me add a similar error message if the cluster doesn't exist, so at least the CLI is similar in behaviour.

Comment (nicolai86) [reply src/lightning_app/cli/cmd_apps.py:30]: hm, implementing this leads to a problem when a cluster is deleted:
we don't delete app releases associated with a cluster, so if I implement a similar validation you reach inconsistent behaviour:

`lightning list apps` will show apps for a cluster that is deleted, but `lightning list apps --cluster-id <deleted-cluster>` will fail because the cluster was deleted.

I'll leave it as it is.

File: src/lightning_app/cli/cmd_apps.py
Comment (nicolai86) [reply src/lightning_app/cli/cmd_apps.py:52]: added other state transitions like "not yet run" and "restarting" :+1:

File: src/lightning_app/cli/cmd_apps.py
Comment (awaelchli) [reply src/lightning_app/cli/cmd_apps.py]: Looks great

**Pull request merged**
